core/object/ref_counted.h

```diff
@@ -181,10 +181,15 @@ class Ref {
 		// do a lot of referencing on references and stuff
 		// mutexes will avoid more crashes?
 
-		if (reference && reference->unreference()) {
-			memdelete(reference);
+		if (reference) {
+			// NOTE: `reinterpret_cast` is "safe" here, as the only way to achieve a non-null
+			// reference is after `ref_pointer` validates the type. This allows for passing
+			// forward-declared types without cascading dependency-chains.
```

Comment:
Given my general note, I'd reword it like this:
```suggestion
			// NOTE: `reinterpret_cast` is "safe" here because we know `T` has simple linear inheritance to `RefCounted`. This guarantees that `T * == `RefCounted *`. 
			Making this assumption allows us to declare `Ref<T>` for forward declared `T` types.
```